pkg/system/names.go

```diff
@@ -16,8 +16,17 @@ limitations under the License.
 
 package system
 
-const (
-	// Namespace holds the K8s namespace where our serving system
+import "os"
+
+var (
+	// Namespace holds the K8s namespace where our build system
 	// components run.
-	Namespace = "knative-serving"
+	Namespace string
 )
+
+func init() {
```

Comment:
What do you think of making this just a getter method rather than having an init which fills in a global (and therefore needs care taken to ensure its called before referencing the global)? I.e. 

```go
const (
  NamespaceEnvKey = "SYSTEM_NAMESPACE"

func Namespace() string {
  return os.Getenv(NamespaceEnvKey)
}
```

Reply:
But if use this method, the default namespace `knative-serving` can't be set. We need the default namespace `knative-serving` in current version.

Reply:
I think @greghaynes was referring to actually adding the defaulting logic into the getter, i.e.:

```golang
const (
  NamespaceEnvKey = "SYSTEM_NAMESPACE"
)

func Namespace() string {
  if ns := os.Getenv(NamespaceEnvKey); ns != "" {
    return ns
  }
  return "knative-serving"
}
```

I think `init()` is [called automatically before running `main`](https://medium.com/golangspec/init-functions-in-go-eac191b3860a) though. Both approaches might be valid in this case, knative-build opted for the init function.

Reply:
+1 on getter.  Why do we need the default?  Tests?

I'd like so see if we can create a shared library for this in knative/pkg to standardize this across Knative.

Reply:
@mattmoor I also hope the knative-serving doesn't need the default namespace and user can free to specify the namespace. But the current knative can't put the serving, build and eventing into the same namespace. The current doc and test case are all specify the knative-serving in `knative-serving` namespace. And also knative-build specify in `knative-build` namespace. So I follow by the code of knative-build to change here.

Reply:
I understand that we can't (yet) put things in the same namespace, but this is one of a number of changes to move in that direction.

I'd like to see us have a shared library for this in `knative/pkg` that folks can pull in to determine the system namespace, assuming it's configured as you have in the yamls.

I'm wondering when the defaulting here actually kicks in (tests?), so that we can take steps to address that.

Reply:
After I check the commit history,  the namespace is hardcode(#1101) in `pkg/names.go` from the beginning. I think this is not only for testing, but also for deployment.
```
// GetServingSystemNamespace returns the namespace where
// serving controllers are deployed to.
func GetServingSystemNamespace() string {
	return "knative-serving-system"
}
```
After #1549, `pkg/names.go` move to `pkg/system/names.go`,  the code has changed as follows:
```
package system

const (
	// Namespace holds the K8s namespace where our serving system
	// components run.
	Namespace = "knative-serving"
)
```

Reply:
I think @mattmoor's question was:

Once you apply your change of reading the namespace from environment (including the YAML changes), is there a place where the default is actually used (hence: do we even need it?).

Your changes make sure that all of our components are properly configured so either there is a place somewhere that isn't configured (and might need to be?) or we don't need the default at all.

Reply:
The default namespaces is unnecessary. The getter method is better than current.  But there will be more code changes. I have updated the code.